### PR TITLE
IMDB Classifier

### LINKED ISSUE #293: Create a text classifier for the IMDB large movie dataset
**Is your feature request related to a problem? Please describe.**
Add a modularized text classifier to Forte.

### PR DESCRIPTION
**Pull request by jrxk**
This PR fixes #293. 

### Description of changes
Added a text classifier for the IMDB large movie dataset based on Texar-PyTorch and BERT. The model expects CSV file inputs with columns (content, label, id).

### Test Conducted
Added an example.

### CODE REVIEW

**Review by hunterhector — CHANGES_REQUESTED** [review]
The main problem is that this classifier didn't use any Forte data processors, how can we apply the data augmentation processors here. By just reading this example, it does not really belong to the Forte repo.

File: forte/models/imdb_text_classifier/data/download_imdb.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import sys
+import subprocess
+
+
+def main():
+    if not os.path.exists("data/IMDB_raw"):
+        subprocess.run("mkdir data/IMDB_raw", shell=True, check=True)
+    # pylint: disable=line-too-long
+    subprocess.run(
+        'wget -P data/IMDB_raw/ https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz',
```
Comment (hunterhector) [inline forte/models/imdb_text_classifier/data/download_imdb.py:26]: don't assume the user environment has any of these command line tools.

**Comment by jrxk** [comment]
> The main problem is that this classifier didn't use any Forte data processors, how can we apply the data augmentation processors here. By just reading this example, it does not really belong to the Forte repo.

I added a pipeline that uses the Forte reader to preprocess data.

File: forte/models/imdb_text_classifier/data/download_imdb.py (hunk 1 above)
Comment (jrxk) [reply forte/models/imdb_text_classifier/data/download_imdb.py:26]: Switched to Forte `maybe_download`

**Comment by jrxk** [comment]
The UDA Experiment PR #320 is based on this PR. Should I create a separate directory for the UDA experiment example, or should I just move the UDA changes to this PR?

**Comment by hunterhector** [comment]
> The UDA Experiment PR #320 is based on this PR. Should I create a separate directory for the UDA experiment example, or should I just move the UDA changes to this PR?

We discuss in our last meeting that we need to have all examples of DA in one folder. Can you coordinate and create that?

**Review by hunterhector — CHANGES_REQUESTED** [review]
I am very puzzled by this PR:
1. Why are we using test set for dev set?
2. Where are the back translation models?

There are a lot more comments inline.

File: forte/models/imdb_text_classifier/utils/data_utils.py [hunk 2]
```diff
@@ -122,7 +125,7 @@ def get_train_examples(self, raw_data_dir):
                            quotechar='"'), "train")
 
     def get_dev_examples(self, raw_data_dir):
-        """See base class."""
+        """The IMDB dataset does not have a dev set so we just use test set"""
```
[the remark this answers is not in the document]
Comment (jrxk) [reply forte/models/imdb_text_classifier/utils/data_utils.py:128]: The IMDB dataset does not have a dev set. It is simply split into 25000 training examples and 25000 test examples.

The user should be aware of this. I can also remove this function if that makes it clearer.

[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/models/imdb_text_classifier/utils/data_utils.py:128]: But when you are doing this experiment, what are we doing with the dev set?

[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/models/imdb_text_classifier/utils/data_utils.py:128]: Do you mean this example is not used? We probably should remove all unused function.

File: forte/models/imdb_text_classifier/utils/data_utils.py [hunk 3]
```diff
@@ -456,9 +475,116 @@ def convert_examples_to_features_and_output_to_files(
...
+
+def back_translation(examples, back_translation_file, data_total_size):
+    """Run back translation."""
+    use_min_length = 10
+    use_max_length_diff_ratio = 0.5
+    logging.info("running bt augmentation")
+
+    text_per_example = 1
+
+    with open(back_translation_file, encoding='utf-8') as inf:
```
[the remark this answers is not in the document]
Comment (jrxk) [reply forte/models/imdb_text_classifier/utils/data_utils.py:536]: This is part of the code for the UDA experiment. I haven't included the UDA code in this PR (yet).

In the UDA experiment, the back translation should be generated by the user and output to a file. They can use their own back translation model or Forte's back translation model.

[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/models/imdb_text_classifier/utils/data_utils.py:536]: I see, do you plan to merge that first? It can be a different PR

File: examples/text_classification/main.py [hunk 4]
```diff
@@ -15,10 +15,11 @@
...
 
-from forte.models.imdb_text_classifier.model import IMDBClassifier
 import config_data
 import config_classifier
 
+from forte.models.imdb_text_classifier.model import IMDBClassifier
+
 
 def main():
     model = IMDBClassifier(config_data, config_classifier)
```
[the remark this answers is not in the document]
Comment (jrxk) [reply examples/text_classification/main.py:25]: The model expects a pickle data format which is specific to the model.

[the remark this answers is not in the document]
Comment (hunterhector) [reply examples/text_classification/main.py:25]: It is not a good idea to put all these into the model. This fix the model so that it can only do one thing.

File: examples/text_classification/main.py [hunk 5]
```diff
@@ -15,10 +15,11 @@
 
 import os
 
-from forte.models.imdb_text_classifier.model import IMDBClassifier
 import config_data
 import config_classifier
 
+from forte.models.imdb_text_classifier.model import IMDBClassifier
```
[the remark this answers is not in the document]
Comment (jrxk) [reply examples/text_classification/main.py:21]: It is in this PR `forte/models/imdb_text_classifier`. It is a BERT text classifier. The BERT model itself is not specific to IMDB but this PR contains preprocessing code specific to IMDB dataset to make it work.

[the remark this answers is not in the document]
Comment (hunterhector) [reply examples/text_classification/main.py:21]: Can you move the preprocessing out from the core model?

File: forte/models/imdb_text_classifier/utils/data_utils.py [hunk 6]
```diff
@@ -74,13 +77,13 @@ def get_labels(self):
         raise NotImplementedError()
 
     @classmethod
-    def _read_tsv(cls, input_file, quotechar=None):
+    def _read_tsv(cls, input_file, quotechar=None):  # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (jrxk) [reply forte/models/imdb_text_classifier/utils/data_utils.py:80]: Our reader reads data from the raw data files (100k separate TXT files) and output to `train.csv` and `test.csv`. Then, our model reads these CSV files and generate pickle files for training, which is why we need `read_tsv` here.

[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/models/imdb_text_classifier/utils/data_utils.py:80]: That's why I am confusing. Why do we still need to go through the whole CSV step? We can directly read from our reader.

File: forte/models/imdb_text_classifier/utils/data_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jrxk) [reply forte/models/imdb_text_classifier/utils/data_utils.py:128]: I can remove this. When training we simply look at the test set result.

**Comment by hunterhector** [comment]
One thing we can do to fix this PR is that @ziqian98 can help resolve many of the Forte parts.

File: forte/models/imdb_text_classifier/config_data.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 # used for bert executor example
 max_batch_tokens = 128
 
-train_batch_size = 32
+train_batch_size = 24
```
[the remark this answers is not in the document]
Comment (jrxk) [reply forte/models/imdb_text_classifier/config_data.py:9]: I put it in the model directory as an example of the expected parameters in `config_data`. The user can simply copy this file if they want to use the model.

[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/models/imdb_text_classifier/config_data.py:9]: We should probably only keep one to reduce maintenance effort.

**Comment by jrxk** [comment]
Closing this PR because we will use Ziqian's classifier as an IMDB classifier example.
